File: src/cf_ip_blacklist/sync.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any

import httpx

from .errors import CloudflareError, SafetyError
from .http import request_with_retry
from .models import CloudflareItem, DesiredList


@dataclass(frozen=True)
class ListDiff:
    additions: tuple[CloudflareItem, ...]
    removals: tuple[str, ...]
    changes: tuple[CloudflareItem, ...]

    @property
    def identical(self) -> bool:
        return not (self.additions or self.removals or self.changes)


def diff_lists(desired: DesiredList, actual: list[CloudflareItem]) -> ListDiff:
    want = {item.ip: item for item in desired.items}
    have = {item.ip: item for item in actual}
    additions = tuple(want[ip] for ip in sorted(set(want) - set(have)))
    removals = tuple(sorted(set(have) - set(want)))
    changes = tuple(
        want[ip] for ip in sorted(set(want) & set(have)) if want[ip].comment != have[ip].comment
    )
    return ListDiff(additions, removals, changes)


class ListsClient:
    def __init__(
        self,
        client: httpx.Client,
        base_url: str,
        account_id: str,
        list_id: str,
        poll_interval: float = 2,
        poll_timeout: float = 120,
    ) -> None:
        self.client, self.base_url = client, base_url.rstrip("/")
        self.account_id, self.list_id = account_id, list_id
        self.poll_interval, self.poll_timeout = poll_interval, poll_timeout
        self.max_retries = 3

    @property
    def items_url(self) -> str:
        return f"{self.base_url}/accounts/{self.account_id}/rules/lists/{self.list_id}/items"

    def get_items(self) -> list[CloudflareItem]:
        response = request_with_retry(self.client, "GET", self.items_url, self.max_retries)
        if response.status_code >= 400:
            raise CloudflareError(f"list read HTTP {response.status_code}")
        try:
            return [
                CloudflareItem.model_validate(
                    {"ip": item["ip"], "comment": item.get("comment", "")}
                )
                for item in response.json().get("result", [])
            ]
        except (TypeError, ValueError) as exc:
            raise CloudflareError("invalid list response") from exc

    def replace(
        self, desired: DesiredList, allow_empty: bool = False, actual_count: int = 0
    ) -> None:
        if actual_count and not desired.items and not allow_empty:
            raise SafetyError("refusing to replace a non-empty remote list with an empty list")
        response = request_with_retry(
            self.client,
            "PUT",
            self.items_url,
            self.max_retries,
            json=[item.model_dump() for item in desired.items],
        )
        if response.status_code >= 400:
            raise CloudflareError(f"list write HTTP {response.status_code}")
        operation_id = response.json().get("result", {}).get("operation_id")
        if operation_id:
            self.wait(operation_id)
        self.verify(desired)

    def verify(self, desired: DesiredList) -> None:
        deadline = time.monotonic() + self.poll_timeout
        while True:
            if diff_lists(desired, self.get_items()).identical:
                return
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise CloudflareError("remote list verification mismatch")
            time.sleep(min(self.poll_interval, remaining))

    def wait(self, operation_id: str) -> None:
        url = (
            f"{self.base_url}/accounts/{self.account_id}/rules/lists/bulk_operations/{operation_id}"
        )
        deadline = time.monotonic() + self.poll_timeout
        while time.monotonic() < deadline:
            response = request_with_retry(self.client, "GET", url, self.max_retries)
            if response.status_code >= 400:
                raise CloudflareError(f"operation poll HTTP {response.status_code}")
            result: dict[str, Any] = response.json().get("result", {})
            status = result.get("status")
            if status in {"completed", "success", "succeeded"}:
                return
            if status in {"failed", "error"}:
                raise CloudflareError(f"Cloudflare operation failed: {result}")
            time.sleep(self.poll_interval)
        raise CloudflareError("Cloudflare operation polling timed out")

File: src/sync.rs
use crate::{
    error::{BlockholeError, Result},
    http::request,
    models::{CloudflareItem, DesiredList, Subject},
};
use reqwest::blocking::Client;
use serde::Deserialize;
use std::{
    collections::{BTreeMap, BTreeSet},
    thread::sleep,
    time::{Duration, Instant},
};
#[derive(Debug, Eq, PartialEq)]
pub struct ListDiff {
    pub additions: Vec<CloudflareItem>,
    pub removals: Vec<Subject>,
    pub changes: Vec<CloudflareItem>,
}
impl ListDiff {
    pub fn identical(&self) -> bool {
        self.additions.is_empty() && self.removals.is_empty() && self.changes.is_empty()
    }
}
pub fn diff(desired: &DesiredList, actual: &[CloudflareItem]) -> ListDiff {
    let want: BTreeMap<_, _> = desired.items.iter().map(|i| (i.ip.clone(), i)).collect();
    let have: BTreeMap<_, _> = actual.iter().map(|i| (i.ip.clone(), i)).collect();
    ListDiff {
        additions: want
            .iter()
            .filter(|(k, _)| !have.contains_key(*k))
            .map(|(_, v)| (*v).clone())
            .collect(),
        removals: have
            .keys()
            .filter(|k| !want.contains_key(*k))
            .cloned()
            .collect(),
        changes: want
            .iter()
            .filter(|(k, v)| have.get(*k).is_some_and(|x| x.comment != v.comment))
            .map(|(_, v)| (*v).clone())
            .collect(),
    }
}
#[derive(Deserialize)]
struct ListResponse {
    result: Vec<ListRaw>,
    result_info: Option<ResultInfo>,
}
#[derive(Deserialize)]
struct ListRaw {
    ip: String,
    #[serde(default)]
    comment: String,
}
#[derive(Deserialize)]
struct ResultInfo {
    cursors: Option<Cursors>,
}
#[derive(Deserialize)]
struct Cursors {
    after: Option<String>,
}
#[derive(Deserialize)]
struct OperationResponse {
    result: Option<Operation>,
}
#[derive(Deserialize)]
struct Operation {
    operation_id: Option<String>,
    status: Option<String>,
}
pub struct ListsClient {
    client: Client,
    base: String,
    account: String,
    list: String,
    retries: usize,
    poll_interval: f64,
    poll_timeout: f64,
}
impl ListsClient {
    pub fn new(
        client: Client,
        base: &str,
        account: &str,
        list: &str,
        retries: usize,
        poll_interval: f64,
        poll_timeout: f64,
    ) -> Self {
        Self {
            client,
            base: base.trim_end_matches('/').into(),
            account: account.into(),
            list: list.into(),
            retries,
            poll_interval,
            poll_timeout,
        }
    }
    fn items_url(&self) -> String {
        format!(
            "{}/accounts/{}/rules/lists/{}/items",
            self.base, self.account, self.list
        )
    }
    pub fn get_items(&self) -> Result<Vec<CloudflareItem>> {
        let mut items = Vec::new();
        let mut cursor = None;
        let mut seen = BTreeSet::new();
        loop {
            let mut url = format!("{}?per_page=500", self.items_url());
            if let Some(ref c) = cursor {
                url.push_str(&format!("&cursor={c}"));
            }
            let response = request(&self.client, reqwest::Method::GET, &url, self.retries, None)?;
            if !response.status().is_success() {
                return Err(BlockholeError::Cloudflare(format!(
                    "list read HTTP {}",
                    response.status()
                )));
            }
            let payload: ListResponse = response.json()?;
            for item in payload.result {
                items.push(CloudflareItem {
                    ip: Subject::parse(&item.ip)?,
                    comment: item.comment,
                });
            }
            let next = payload
                .result_info
                .and_then(|i| i.cursors)
                .and_then(|c| c.after);
            match next {
                None => return Ok(items),
                Some(c) if !seen.insert(c.clone()) => {
                    return Err(BlockholeError::Cloudflare(
                        "list response pagination cursor repeated".into(),
                    ));
                }
                Some(c) => cursor = Some(c),
            }
        }
    }
    pub fn replace(
        &self,
        desired: &DesiredList,
        actual_count: usize,
        allow_empty: bool,
    ) -> Result<()> {
        if actual_count > 0 && desired.items.is_empty() && !allow_empty {
            return Err(BlockholeError::Safety(
                "refusing to replace a non-empty remote list with an empty list".into(),
            ));
        }
        let body = serde_json::to_value(&desired.items)?;
        let response = request(
            &self.client,
            reqwest::Method::PUT,
            &self.items_url(),
            self.retries,
            Some(body),
        )?;
        if !response.status().is_success() {
            return Err(BlockholeError::Cloudflare(format!(
                "list write HTTP {}",
                response.status()
            )));
        }
        let operation: OperationResponse = response.json()?;
        if let Some(id) = operation.result.and_then(|r| r.operation_id) {
            self.wait(&id)?;
        }
        let deadline = Instant::now() + Duration::from_secs_f64(self.poll_timeout);
        loop {
            if diff(desired, &self.get_items()?).identical() {
                return Ok(());
            }
            if Instant::now() >= deadline {
                return Err(BlockholeError::Cloudflare(
                    "remote list verification mismatch".into(),
                ));
            }
            sleep(Duration::from_secs_f64(self.poll_interval));
        }
    }
    fn wait(&self, id: &str) -> Result<()> {
        let url = format!(
            "{}/accounts/{}/rules/lists/bulk_operations/{id}",
            self.base, self.account
        );
        let deadline = Instant::now() + Duration::from_secs_f64(self.poll_timeout);
        loop {
            let response = request(&self.client, reqwest::Method::GET, &url, self.retries, None)?;
            if !response.status().is_success() {
                return Err(BlockholeError::Cloudflare(format!(
                    "operation poll HTTP {}",
                    response.status()
                )));
            }
            let payload: OperationResponse = response.json()?;
            match payload.result.and_then(|r| r.status) {
                Some(status)
                    if ["completed", "success", "succeeded"].contains(&status.as_str()) =>
                {
                    return Ok(());
                }
                Some(status) if ["failed", "error"].contains(&status.as_str()) => {
                    return Err(BlockholeError::Cloudflare(format!(
                        "Cloudflare operation failed: {status}"
                    )));
                }
                _ => {}
            }
            if Instant::now() >= deadline {
                return Err(BlockholeError::Cloudflare(
                    "Cloudflare operation polling timed out".into(),
                ));
            }
            sleep(Duration::from_secs_f64(self.poll_interval));
        }
    }
}
